File: src/engagement_scoring.R
# Load necessary libraries
library(dplyr)

source("src/utils.R")

# Function to score customer engagement
score_engagement <- function(data) {
  require_columns(data, c("total_interactions", "total_page_views", "total_purchases", "avg_purchase_value"))
  data %>%
    mutate(
      engagement_score = total_interactions +
        total_page_views +
        safe_divide(total_purchases, avg_purchase_value, default = 0)
    )
}

run_engagement_scoring <- function(
  features_path = "data/features.csv",
  output_path = "data/engagement_scores.csv"
) {
  data <- read.csv(features_path)

  engagement_scores <- score_engagement(data)

  write.csv(engagement_scores, output_path, row.names = FALSE)
  engagement_scores
}

if (sys.nframe() == 0) {
  run_engagement_scoring()
}

File: src_py/engagement_scoring.py
from __future__ import annotations

from pathlib import Path

from src_py.utils import PROJECT_ROOT, read_csv, safe_divide, write_csv


def score_engagement(features_path: Path, output_path: Path) -> None:
    rows = read_csv(features_path)
    scored = []
    for row in rows:
        total_interactions = float(row["total_interactions"])
        total_page_views = float(row["total_page_views"])
        total_purchases = float(row["total_purchases"])
        avg_purchase_value = float(row["avg_purchase_value"])

        engagement_score = total_interactions + total_page_views + safe_divide(
            total_purchases, avg_purchase_value, default=0
        )

        updated = dict(row)
        updated["engagement_score"] = round(engagement_score, 2)
        scored.append(updated)

    fieldnames = list(scored[0].keys()) if scored else []
    write_csv(output_path, scored, fieldnames)


def run() -> None:
    data_dir = PROJECT_ROOT / "data"
    score_engagement(data_dir / "features.csv", data_dir / "engagement_scores.csv")


if __name__ == "__main__":
    run()
